File: javascript/Array/levelUpYourCoreProgrammingSkills/missingNumber/missingNumber.js
// Find the missing number from an array of integers.
// The array is expected to have all the integers from 0 up to the arrays length but one is missing.
// Write a function that takes only O(N log N)

// Example:
// Input: [5, 2, 4, 1, 0]
// Output: 3

// Input: [9, 3, 2, 5, 6, 7, 1, 0, 4]
// Output: 8

const missingNumber = function (inputArray) {
  inputArray.sort((a, b) => a - b)
  for (let i = 0; i < inputArray.length; i += 1) {
    if (inputArray[i] != i) {
      return i
    }
  }
  return "No missing number"
}

console.log(missingNumber([5, 2, 4, 1, 0]))
console.log(missingNumber([9, 3, 2, 5, 6, 7, 1, 0, 4]))
console.log(missingNumber([9, 3, 2, 5, 6, 7, 1, 0, 4, 8]))

File: ruby/missing_number/missing_number.rb
# You're writing a function that accepts an array of distinct integers from 0, 1, 2, 3, up to N.
# However, the array will be missing one integer, and your functiong is to return the missing one.

# Example:
# Input:
# [2, 3, 0, 6, 1, 5]
# Output:
# 4

# Input:
# [8, 2, 3, 9, 4, 7, 5, 0, 6]
# Output:
# 1

def missing_number(input_array)
  number_hash = {}

  i = 0
  while i < input_array.length
    if !number_hash[input_array[i]]
      number_hash[input_array[i]] = true
    end
    i += 1
  end
  number_hash
end

p missing_number([8, 2, 3, 9, 4, 7, 5, 0, 6])
p missing_number([2, 3, 0, 6, 1, 5])